src/main/kotlin/cacophony/regex/RegexParser.kt

```diff
@@ -0,0 +1,148 @@
+package cacophony.regex
+
+import cacophony.utils.AlgebraicRegex
+import java.util.EmptyStackException
+import java.util.Stack
+
+fun parseRegex(str: String): AlgebraicRegex {
+    val parser = RegexParser()
+    val regex = "($str)"
+    var specialCharacter = false
+    var lastEscaped = -1
+    regex.indices.forEach {
+        val c = regex[it]
+        if (specialCharacter) {
+            if (c !in SPECIAL_CHARACTER_MAP.keys) throw RegexSyntaxErrorException()
+            if (it > 1 && (regex[it - 2] !in "(|" || lastEscaped == it - 2)) parser.pushOperation(ConcatOperator)
+            parser.pushRegex(SPECIAL_CHARACTER_MAP[c]!!) // safe because of previous check
+            specialCharacter = false
+            lastEscaped = it
+        } else {
+            when (c) {
+                '(' -> {
+                    if (it != 0 && (regex[it - 1] !in "(|" || lastEscaped == it - 1)) parser.pushOperation(ConcatOperator)
+                    parser.pushOperation(LeftParenthesis)
+                }
+                ')' -> parser.pushOperation(RightParenthesis)
+                '*' -> parser.pushOperation(StarOperator)
+                '|' -> parser.pushOperation(UnionOperator)
+                '\\' -> specialCharacter = true
+                else -> {
+                    if (regex[it - 1] !in "(|") parser.pushOperation(ConcatOperator)
+                    parser.pushSymbol(c)
+                }
+            }
+        }
+    }
+    return parser.finalize().toAlgebraicRegex()
+}
+
+private class RegexParser {
+    val result = Stack<RegexType>()
+    val stack = Stack<StackOperator>()
+
+    fun pushOperation(operator: Operator) {
+        try {
+            operator.addToStack(this)
+        } catch (e: EmptyStackException) {
+            throw RegexSyntaxErrorException()
+        }
+    }
+
+    fun pushSymbol(c: Char) {
+        result.push(Atom(c))
+    }
+
+    fun pushRegex(c: RegexType) {
+        result.push(c)
+    }
+
+    fun finalize(): RegexType {
+        if (result.size != 1 || stack.isNotEmpty()) throw RegexSyntaxErrorException()
+        return result.pop()
+    }
+}
+
+private sealed class Operator {
+    abstract fun addToStack(parser: RegexParser)
+}
```

Comment:
Maybe I don't understand, but why is there Operator and StackOperator?

Reply:
Operator is anything that acts as operator in regex syntax (implicite concat, |, (, ) and *. However, ) and * never actually land on stack of operators in parser (bacause they are resolved immediately), therefore do not need priority. StackOperators can land on the stack, therefore need they priority for additional logic

Reply:
So I think it would be nice to just have Operator with priority, this priority will be just anything for ) and *. Code would be cleaner I think

Reply:
Having ) as something which can be on stack would require handling it (probably by throwing) in every type matching on top stack element, this way we have static guarantees. Removed this class anyway, Star can be now on stack, right parenthesis is handled separately (I guess this way is also correct by static inspection? Dunno if cleaner)